### PR TITLE
Support the same dictation formatters across both modal and modeless dictation.

### PR DESCRIPTION
**Pull request by wolfmanstout**
These are the changes we discussed in last week's knausj grooming meeting. Happy to discuss again now that folks have a simplified diff to look at:

1. Add a 'capitalise next word' type state to dictation mode with the 'cap' command along with 'no cap' to do the opposite. Also a command to skip any preceding space for the next word. This is implemented as both Talon commands and also a post processor on the 'free text' command <user.raw_prose>: auto_insert(raw_prose).
2. Make it so the 'sentence', 'speak', and 'say' commands in command mode can do this kind of inline capitalisation/space toggling. So you could say "sentence this is my friend cap paul full stop" and get "This is my friend Paul."
3. Fix spacing around quotes and smiley faces by inserting and then straightening curly quotes.
4. Corresponding to (1), add new "cap that"-style commands to reformat the most recent utterance.

### CODE REVIEW

File: code/dictation.py
Comment (wolfmanstout) [inline code/dictation.py:181]: Note that this is a behavior change in dictation mode, which now no longer starts with a capitalized letter, consistent with modeless "speak" command. We could easily avoid this if desired.

File: code/dictation.py
Comment (wolfmanstout) [inline code/dictation.py:20]: Could use existing vocab features except I think they are an awkward fit for whitespace, and there is code here that interacts with these.

File: code/dictation.py
Comment (wolfmanstout) [inline code/dictation.py:68]: We could do away with this, but I've found it more intuitive.

File: code/dictation.py
Comment (wolfmanstout) [inline code/dictation.py:263]: This change could be dropped without affecting other pieces. I found this made formatters easier to use within dictation because you still get the auto-spacing.

File: code/dictation.py
Comment (rntz) [inline code/dictation.py:74]: This is orthogonal to this PR, but I think we could get away with setting `words = [item]` here. The auto-capitalization code doesn't care about begin handed words one-by-one, so splitting words out is not really necessary. This would automatically preserve whitespace instead of needing to have a special-case for it.

**Review by rntz — CHANGES_REQUESTED** [review]
Cool stuff! I have a few specific concerns as well as some questions, but I'm looking forward to getting this merged and being able to use formatting commands more freely.

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py:20]: Suppose you freshly enter dictation mode and you need to start with a space. This lets you add that. Conceptually, it is the opposite of "no space".

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py:24]: Hmm, I thought I ran into problems putting non-key characters into that file, but I can't reproduce it. Personally, I only want these present in dictation so I can reformat them, and there currently is no dictation-only list in keys.py (which makes sense given it is keys.py). I'm fine maintaining this difference in my fork, though, if this is what you all would like.

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py:46]: Yes, that's exactly why I'm doing it. I agree this is non-obvious and I'm happy to document if we decide to keep this in. Personally I don't want to output unstraightened quotes -- I can't do this with my keyboard so I'm fine not having that with voice, and it can cause trouble if they make it into code. But I would like to make this work for folks like yourself if you find it useful. I'm not sure there is a both clean and simple way to do this, although we can use any characters (or sequence of characters) we want, so we could do double curly quotes or something like that (we'd need to fix the regexes).

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py:73]: Agreed -- I kept this as is since I was already changing so much and I didn't notice any problems with the current implementation. Happy to make this change here if you would like.

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py:91]: I added space so that # is aligned.

File: modes/dictation_mode.talon
Comment (wolfmanstout) [reply modes/dictation_mode.talon]: Yep -- Dragon thing. Added comment.

File: modes/dictation_mode.talon
Comment (wolfmanstout) [reply modes/dictation_mode.talon:6]: Correct. Another benefit of this change is you can pause after modifiers like "cap", which is in keeping with modal dictation being the pause-friendly complement to modeless dictation.

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py:90]: I added "apostrophe ess": "'s" to keys.py and that only formats correctly if I make this change. How do you do achieve this?

File: modes/dictation_mode.talon
Comment (wolfmanstout) [inline modes/dictation_mode.talon:11]: FYI I added anchors to these commands. I don't need this in my grammar because I'm using the experimental anchor-file feature so I just move these to a separate file so they can't be chained with dictation.

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py]: Done and changed to reset_context.

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py]: Sure, done. Personally I think it's a bit of a toss up between a couple booleans vs. introducing string matches that aren't (runtime) type checked.

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py]: Good idea, done.

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py]: Originally I did in fact reuse self.state. The problem is it's nice to override default capitalization (e.g. make proper names lowercase) when desired, so I needed the ability to _force_, which is more than self.state offered. In terms of separation of concerns, I think it's cleaner to have separate state instance variables instead of trying to package everything in one. I agree there is some duplication between format_first_letter and auto_capitalize, though, although I'm not sure trying to do all that in one function is going to simplify things.

**Review by wolfmanstout — COMMENTED** [review]
I think I've answered all your questions. Thanks for the thorough review!

File: code/dictation.py
Comment (rntz) [reply code/dictation.py:90]: Ah, I don't have that feature, but I can see how it's useful. The reason I'm not sure about making `'` unconditionally require no space before it is because sometimes it's used for quotation instead of `"` (eg. sometimes in British English this is the preferred style). Perhaps we should special-case possessives by adding `'s\b` to no_space_before. Does that seem like a good approach?

File: code/dictation.py
Comment (rntz) [reply code/dictation.py:46]: I was thinking we could bind "open quote" to `' "'` (double quote with a space before it) and close quote to `'" '` (double quote with a space after) work? Unfortunately, this produces spurious spaces at the beginning/end of documents.

I'll have to think more about this problem. But I think your use-case is more common/important than mine - most people want straight quotes and not curly. So if push comes to shove I'm inclined to accept your approach on this; we can always fix it up with another PR later.

File: code/dictation.py
Comment (rntz) [reply code/dictation.py]: Yeah, I agree.

**Comment by rntz** [comment]
by the way, this is unrelated to the PR, but you might be interested in trying out context-sensitive dictation, which looks at the characters around the cursor to determine how to space/autocapitalize before inserting during dictation mode. It might reduce the need for commands like "no space", etc. You can enable it by editing settings.talon, https://github.com/knausj85/knausj_talon/blob/master/settings.talon#L51.

It's off by default because it works by select-and-copying stuff around the cursor, which may not work consistently in all applications. Would appreciate feedback if it doesn't work well for you - maybe with enough polish we could enable it by default.

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py]: Sounds good to me; I just made the change.

File: modes/dictation_mode.talon
Comment (wolfmanstout) [reply modes/dictation_mode.talon]: And yes works fine under Conformer too.

**Review by wolfmanstout — COMMENTED** [review]
Regarding your note about context sensitive dictation, I do in fact have that enabled and really like the feature! I still use the changes here particularly in modeless dictation commands. Also there is a nasty bug on WSLg that prevents me from being able to use context sensitive dictation with Emacs without causing it to freeze.

I have bumped into other issues with context sensitive dictation, particularly having to do with swallowed keystrokes and the clipboard behaving slowly, which I have currently worked around by adding delays to selected_text() in certain apps. I'm not sure if there's a better way.

File: code/dictation.py
Comment (phillco) [reply code/dictation.py:181]: Could we make this a configuration option? I don't see need to change the default behavior, but that could make it easy for people to pick what they like.

Comment (wolfmanstout) [reply code/dictation.py:181]: That would certainly work for me. I imagine we'd would want to preserve the "speak" behavior instead of aligning it. That gives us two options for implementation: either change what DictationFormat.reset does and special-case "speak" to set state to None, or update calls to DictationFormat.reset, perhaps by having them call user.dictation_format_reset() instead and putting this behavior in there. I'm weakly inclined towards the latter.

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py:46]: We could add a "straighten_quotes" setting and check that in the two places that straightening is done in this file. This also lets people set it per-app, which I could imagine some folks wanting.

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py:181]: Actually what we'd do in the former case is set formatter.state = None in apply_formatting() here, which I'm already doing. That's probably the simpler approach then.

**Comment by wolfmanstout** [comment]
@rntz another quick note on context sensitive dictation: I think a nice optimization would be to avoid this check in an utterance with a sequence of dictation commands for every command after the first. Obviously only safe to do if the commands don't include any movements, so it may be tricky to implement.

File: code/dictation.py
Comment (rntz) [reply code/dictation.py:46]: I think it's better to leave the PR as-is rather than adding a setting. I have an idea for how to get the spacing right without using curly quotes and then straightening them, which would obviate the need for a setting. I'll make an issue for it shortly.

**Review by rntz — APPROVED** [review]
Okay, I think this is basically good to go. However, before we merge this, I'd like at least one other maintainer to approve it. In particular, there are a few breaking changes:

1. Dictation mode would no longer start with a capital letter. I have no strong feelings on this. If anyone prefers the current behavior, though, we can shove this change to a separate PR/issue for discussion.

2. This PR changes the behavior of `dictation_insert_raw` and thus dictation-mode formatters (https://github.com/knausj85/knausj_talon/blob/master/modes/dictation_mode.talon#L60) to add auto-spacing around them. I think this is an improvement, but if anyone strongly prefers current behavior, we can shove this change to a separate PR.

3. Curly quotes `“`/`”` will get straightened to `"` inside dictated text. This should only change behavior if you've mapped commands/keys to insert curly quotes. I think this is OK for now and I have a plan to fix this post-merge, but if anyone strongly objects to this behavior, we can block this PR until this issue is addressed.

File: code/dictation.py
Comment (rntz) [reply code/dictation.py:181]: @phillco Do you prefer the current behavior?

**Comment by rntz** [comment]
@splondike I'd appreciate your review, especially with respect to the breaking changes I mentioned above. (Or a review from any maintainer, but I had to pick someone to request a review from :P.)

**Comment by rntz** [comment]
Some post-merge action items:

1. Add comments explaining why we're replacing curly quotes by straight quotes and how this makes the `(open|close) quote` snippets work. At least comment `dictation_insert`; might also want to comment the `prose` capture and the `prose_snippets` dictionary.

2. In `apply_formatting`, simplify `words = item.split(" ") if item != " " else [" "]` to be just `words = [item]`. Test that this works.

3. Fix the straightening-curly-quotes problem while still allowing properly-spaced `"` via `(open|close) quote` commands. This needs its own PR. I see two ways to do this. Option (a) is to map `(open|close) quote` to non-string values (maybe an Enum) that gets handled specially by `DictationFormat.format`. Option (b) is to map `open quote` to `' "'` and `close quote` to `'" '` and then treat leading/trailing spaces as "soft" in `DictationFormat.format`, stripping them if they're unnecessary. I mildly prefer (a).

(1-2) could be done pre-merge if @wolfmanstout feels like it but I don't think they should block merging and I'm happy to handle them myself post-merge.

**Comment by wolfmanstout** [comment]
This plan sounds good. I'm happy to do the little cleanup bits you mention (and anything else that comes out of a final review). I understand this is a larger change and we need to be careful particularly with breaking changes.

File: code/dictation.py
Comment (pcohensc) [reply code/dictation.py:181]: I do, although it's easy enough for me to change it in my fork.

Comment (phillco) [reply code/dictation.py:181]: I do, although it's easy enough for me to change it in my fork.

**Comment by wolfmanstout** [comment]
Applied (1) and (2). As for (3), I don't think (a) would trivially work, because the curly quotes (intentionally) make it into the contextual history to affect future utterances, and this is currently a string. I suppose (b) could work, except that the stripping logic will have to deal with several edge cases including the start of form and the new force no-space functionality, and it shouldn't break in cases where the user explicitly adds a space (i.e. using "spacebar").

**Comment by splondike** [comment]
I admit I'm having a little difficulty reviewing the PR because I haven't looked in this code section before. My understanding is this PR primarily intends to do the following:

1. Add a 'capitalise next word' type state to dictation mode with the 'cap' command along with 'no cap' to do the opposite. Also a command to skip any preceding space for the next word. This is implemented as both Talon commands and also a post processor on the 'free text' command `<user.raw_prose>: auto_insert(raw_prose)`.
2. Make it so the 'sentence', 'speak', and 'say' commands in command mode can do this kind of inline capitalisation/space toggling. So you could say "sentence this is my friend cap paul full stop" and get "This is my friend Paul."

For 1. I'm not super sold on the UX of having the extra persistent state (force_no_space, force_capitalization) over doing more of the existing 'cap <user.word>' type stuff (so we'd add 'no cap <user.word>'. The dictation_format singleton is already doing some persistent state I suppose. I guess we have different intuitions here.

For 2. this is fine I think, I guess it's a bit of an admission that we can't do everything we need with Talon's built in command chaining. Ideally we'd be able to express this desire in Talon matchers rather than having a second parser layer. Oh well.

I've added a couple of comments, in particular I think the change to the text capture is likely to break things (though I haven't gone looking beyond how it breaks the escape command), so I would suggest not doing it. Also be aware that you're able to write unit tests against these files (see https://github.com/knausj85/knausj_talon#automated-tests) if that seems useful for development/maintenance.

**Comment by wolfmanstout** [comment]
@splondike I'm having trouble getting automated tests to run; it says that no tests are found. I'm guessing this is related to:
https://github.com/knausj85/knausj_talon/blob/9c622539cd0d0af35a8d87cc159b8792ceb45754/tests/test_formatters.py#L3

How do I enable talon.test_mode?

**Comment by rntz** [comment]
Just a note that #641 got merged, so this PR might need to be updated. The only necessary change should be to replace calls to `actions.dictate.replace_words` with `actions.user.replace_phrases`.

**Comment by splondike** [comment]
:facepalm: Forgot to publish my inline comments, that's done now.

I think you've convinced me about the main UX considerations, so would just have to address those and fix the merge conflict and it's good from my perspective.

I think the issue you've got with pytest is you're using the Talon Python venv. You need to use a non Talon Python runtime (with pytest installed). The test_mode thing comes from the fact that I've stubbed out the Talon API in tests/stubs/ and configured the Python module loader to use that in conftest.py. I should update the README to cover both those since it's not really clear what's happening at the moment.

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py]: I agree completely, this was overeager code reuse. I was thinking of this as providing a superset of the functionality, but you are right that it interacts with words like cap. I brought the old functions back.

File: modes/dictation_mode.talon
Comment (wolfmanstout) [reply modes/dictation_mode.talon:12]: This comes from Dragon, but I agree there's no reason the inconsistency should live on. I'm now using "cap" consistently everywhere, except when Dragon requires otherwise. It simply refuses to hear "no cap" and will bias to its own "no-caps" (which it prints as "no cap"), so I need to add that or it doesn't work. Tested both modal and modeless versions with Dragon and Conformer.

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py]: I did slightly different syntax (with style more in keeping with the other regexes here) but the effect should be equivalent.

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py:281]: @splondike Note this is a breaking change, I can revert if you'd like. It's not used in many places, though.

File: code/dictation.py
Comment (wolfmanstout) [reply code/dictation.py]: Reverted.

File: code/dictation.py
Comment (wolfmanstout) [inline code/dictation.py:71]: Note I changed this to be in consistent with @rntz suggestion and the behavior below.

**Comment by wolfmanstout** [comment]
Added unit tests. I think this is ready to be merged.

**Pull request merged**
